presto-main/src/main/java/com/facebook/presto/operator/MergeJoinRightPages.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.common.Page;
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.type.Type;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class MergeJoinRightPages
+{
+    private int positionCount;
+
+    private final List<Type> types;
+    private final List<Integer> joinChannels;
+    private final List<Integer> outputChannels;
+    private List<Page> pages;
+    private boolean complete;
+
+    private Page singleValueMatchPage;
+
+    public MergeJoinRightPages(List<Type> types, List<Integer> joinChannels, List<Integer> outputChannels)
+    {
+        this.types = types;
+        this.joinChannels = joinChannels;
+        this.outputChannels = outputChannels;
+        this.positionCount = 0;
+        this.pages = new ArrayList<>();
+    }
+
+    public void startMatch(Page page, int position)
+    {
+        clear();
+        singleValueMatchPage = page.getSingleValuePage(position);
+    }
+
+    public void addPage(Page page, int startPosition, int endPosition)
+    {
+        if (endPosition < page.getPositionCount()) {
+            complete = true;
+        }
+
+        Block[] blocks = new Block[outputChannels.size()];
+        for (int i = 0; i < outputChannels.size(); i++) {
+            Block block = page.getBlock(outputChannels.get(i));
+            block = block.copyRegion(startPosition, endPosition - startPosition);
+            blocks[i] = block;
+        }
+        pages.add(new Page(endPosition - startPosition, blocks));
```

Comment:
Refer to `PagesIndex` `computeNext`
1. We only start to build pages later when it's used
2. we can merge several data chunk into one page instead of creating one page per data chunk